metron-platform/metron-writer/src/main/java/org/apache/metron/writer/kafka/KafkaWriter.java

```diff
@@ -197,6 +209,15 @@ public void init(Map stormConf, TopologyContext topologyContext, WriterConfigura
     return producerConfig;
   }
 
+  public Optional<String> getKafkaTopic(JSONObject message) {
```

Comment:
I am assuming this logic applies to any topology that uses a `KafkaWriter`.  Would it be easy enough for a user to run into an infinite loop scenario if they have any two sequential topologies both using a `KafkaWriter`?

Parser -> Stage1 -> Stage2 -> ...

* In the Parser, it ingests a message where the "outputTopic" = "stage1".  
* This sends the message to my Stage1 processing
* If the Stage1 logic does not change the value of that field for whatever reason, then the message will go right back to Stage1 and be reprocessed.
* Wash, rinse, repeat and you've got a mess on your hands that is difficult to debug.

Maybe I am thinking too hard about this.  There may be nothing we can really do about that.  With power comes responsibility.  If an advanced user wants to customize routing, then they need to own this risk.

Reply:
They absolutely could create loops in kafka, which wouldn't be ideal.  I don't know how I could prevent it, sadly.  I think this falls in the "with great power comes great responsibility".  Should we call this a spiderman bug (spiderbug?).

Reply:
What if we unset the value of the topic field after the redirect occurs?  This would force the user to set the field again, if they really want another redirect.

It seems like on day 1 of a user attempting to use this functionality, they are going to fall into this trap.  To really use this functionality, you need to set the value in one step and then unset or change it in the next step.  If you don't, its going to loop.

Reply:
It's unclear to me that we really want to do that.  It could be that users are going to have logic that depends on that field downstream.  It seems wrong to me to remove a field that they're adding or have in the message.  

I mean, in order to create a loop, one of two things would have to happen:
* the user specifies the input topic as the output topic (aka a simple loop), which removing the kafka topic field wouldn't help because it is likely to be computed (e.g. in the manual test script)
* the user creates a non-simple loop where by sensor A -> B -> ... C -> A, but it's exceedingly unlikely that they're all going to be parsers of the same type, so messages from C will unlikely fail to parse in A.  In the case that they do, it's likely that the kafka topic will be computed in first parser, so it'll be recomputed and removing the field after the first parse won't have helped.

TL;DR
Ultimately, I think removing the field won't appreciably help the situation and puts us into the state of removing data, which makes me uncomfortable.

Reply:
I think I see your point.  I think we're ok, though, because we're not defaulting `kafka.topicField`.  If it's not specified AND the `kafka.topic` isn't specified, then we don't send the message anywhere.  I did this specifically so people didn't accidentally forget to unset a field and end up in a loop.  You have to go out of your way (and set the `kafka.topicField`) to make the mistake.

Reply:
Ok, that makes sense to me.  A user would have to take additional steps to really shoot their self in the foot here.  If 'topic field' was a global setting this might happen, but its not.  

I agree with you.  I don't think there is anything we need to change here.